File: mediapipe/python/solutions/hands.py
# Copyright 2020 The MediaPipe Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""MediaPipe Hands."""

import enum
from typing import NamedTuple

import numpy as np

# pylint: disable=unused-import
from mediapipe.calculators.core import constant_side_packet_calculator_pb2
from mediapipe.calculators.core import gate_calculator_pb2
from mediapipe.calculators.core import split_vector_calculator_pb2
from mediapipe.calculators.tensor import image_to_tensor_calculator_pb2
from mediapipe.calculators.tensor import inference_calculator_pb2
from mediapipe.calculators.tensor import tensors_to_classification_calculator_pb2
from mediapipe.calculators.tensor import tensors_to_detections_calculator_pb2
from mediapipe.calculators.tensor import tensors_to_landmarks_calculator_pb2
from mediapipe.calculators.tflite import ssd_anchors_calculator_pb2
from mediapipe.calculators.util import association_calculator_pb2
from mediapipe.calculators.util import detections_to_rects_calculator_pb2
from mediapipe.calculators.util import logic_calculator_pb2
from mediapipe.calculators.util import non_max_suppression_calculator_pb2
from mediapipe.calculators.util import rect_transformation_calculator_pb2
from mediapipe.calculators.util import thresholding_calculator_pb2
# pylint: enable=unused-import
from mediapipe.python.solution_base import SolutionBase
# pylint: disable=unused-import
from mediapipe.python.solutions.hands_connections import HAND_CONNECTIONS
# pylint: enable=unused-import


class HandLandmark(enum.IntEnum):
  """The 21 hand landmarks."""
  WRIST = 0
  THUMB_CMC = 1
  THUMB_MCP = 2
  THUMB_IP = 3
  THUMB_TIP = 4
  INDEX_FINGER_MCP = 5
  INDEX_FINGER_PIP = 6
  INDEX_FINGER_DIP = 7
  INDEX_FINGER_TIP = 8
  MIDDLE_FINGER_MCP = 9
  MIDDLE_FINGER_PIP = 10
  MIDDLE_FINGER_DIP = 11
  MIDDLE_FINGER_TIP = 12
  RING_FINGER_MCP = 13
  RING_FINGER_PIP = 14
  RING_FINGER_DIP = 15
  RING_FINGER_TIP = 16
  PINKY_MCP = 17
  PINKY_PIP = 18
  PINKY_DIP = 19
  PINKY_TIP = 20


_BINARYPB_FILE_PATH = 'mediapipe/modules/hand_landmark/hand_landmark_tracking_cpu.binarypb'


class Hands(SolutionBase):
  """MediaPipe Hands.

  MediaPipe Hands processes an RGB image and returns the hand landmarks and
  handedness (left v.s. right hand) of each detected hand.

  Note that it determines handedness assuming the input image is mirrored,
  i.e., taken with a front-facing/selfie camera (
  https://en.wikipedia.org/wiki/Front-facing_camera) with images flipped
  horizontally. If that is not the case, use, for instance, cv2.flip(image, 1)
  to flip the image first for a correct handedness output.

  Please refer to https://solutions.mediapipe.dev/hands#python-solution-api for
  usage examples.
  """

  def __init__(self,
               static_image_mode=False,
               max_num_hands=2,
               model_complexity=1,
               min_detection_confidence=0.5,
               min_tracking_confidence=0.5):
    """Initializes a MediaPipe Hand object.

    Args:
      static_image_mode: Whether to treat the input images as a batch of static
        and possibly unrelated images, or a video stream. See details in
        https://solutions.mediapipe.dev/hands#static_image_mode.
      max_num_hands: Maximum number of hands to detect. See details in
        https://solutions.mediapipe.dev/hands#max_num_hands.
      model_complexity: Complexity of the hand landmark model: 0 or 1.
        Landmark accuracy as well as inference latency generally go up with the
        model complexity. See details in
        https://solutions.mediapipe.dev/hands#model_complexity.
      min_detection_confidence: Minimum confidence value ([0.0, 1.0]) for hand
        detection to be considered successful. See details in
        https://solutions.mediapipe.dev/hands#min_detection_confidence.
      min_tracking_confidence: Minimum confidence value ([0.0, 1.0]) for the
        hand landmarks to be considered tracked successfully. See details in
        https://solutions.mediapipe.dev/hands#min_tracking_confidence.
    """
    super().__init__(
        binary_graph_path=_BINARYPB_FILE_PATH,
        side_inputs={
            'model_complexity': model_complexity,
            'num_hands': max_num_hands,
            'use_prev_landmarks': not static_image_mode,
        },
        calculator_params={
            'palmdetectioncpu__TensorsToDetectionsCalculator.min_score_thresh':
                min_detection_confidence,
            'handlandmarkcpu__ThresholdingCalculator.threshold':
                min_tracking_confidence,
        },
        outputs=[
            'multi_hand_landmarks', 'multi_hand_world_landmarks',
            'multi_handedness'
        ])

  def process(self, image: np.ndarray) -> NamedTuple:
    """Processes an RGB image and returns the hand landmarks and handedness of each detected hand.

    Args:
      image: An RGB image represented as a numpy ndarray.

    Raises:
      RuntimeError: If the underlying graph throws any error.
      ValueError: If the input image is not three channel RGB.

    Returns:
      A NamedTuple object with the following fields:
        1) a "multi_hand_landmarks" field that contains the hand landmarks on
           each detected hand.
        2) a "multi_hand_world_landmarks" field that contains the hand landmarks
           on each detected hand in real-world 3D coordinates that are in meters
           with the origin at the hand's approximate geometric center.
        3) a "multi_handedness" field that contains the handedness (left v.s.
           right hand) of the detected hand.
    """

    return super().process(input_data={'image': image})

File: mediapipe/java/com/google/mediapipe/solutions/hands/Hands.java
// Copyright 2021 The MediaPipe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.mediapipe.solutions.hands;

import android.content.Context;
import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.mediapipe.formats.proto.LandmarkProto.LandmarkList;
import com.google.mediapipe.formats.proto.LandmarkProto.NormalizedLandmarkList;
import com.google.mediapipe.formats.proto.ClassificationProto.Classification;
import com.google.mediapipe.formats.proto.ClassificationProto.ClassificationList;
import com.google.mediapipe.framework.MediaPipeException;
import com.google.mediapipe.framework.Packet;
import com.google.mediapipe.solutioncore.ErrorListener;
import com.google.mediapipe.solutioncore.ImageSolutionBase;
import com.google.mediapipe.solutioncore.OutputHandler;
import com.google.mediapipe.solutioncore.ResultListener;
import com.google.mediapipe.solutioncore.SolutionInfo;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;

/**
 * MediaPipe Hands Solution API.
 *
 * <p>MediaPipe Hands processes a {@link TextureFrame} or a {@link Bitmap} and returns the hand
 * landmarks and handedness (left v.s. right hand) of each detected hand. Please refer to
 * https://solutions.mediapipe.dev/hands#android-solution-api for usage examples.
 */
public class Hands extends ImageSolutionBase {
  private static final String TAG = "Hands";

  /** Value class representing hand connection. */
  @AutoValue
  public abstract static class Connection {
    static Connection create(int start, int end) {
      return new AutoValue_Hands_Connection(start, end);
    }

    public abstract int start();

    public abstract int end();
  }

  public static final ImmutableSet<Connection> HAND_CONNECTIONS =
      ImmutableSet.of(
          Connection.create(HandLandmark.WRIST, HandLandmark.THUMB_CMC),
          Connection.create(HandLandmark.THUMB_CMC, HandLandmark.THUMB_MCP),
          Connection.create(HandLandmark.THUMB_MCP, HandLandmark.THUMB_IP),
          Connection.create(HandLandmark.THUMB_IP, HandLandmark.THUMB_TIP),
          Connection.create(HandLandmark.WRIST, HandLandmark.INDEX_FINGER_MCP),
          Connection.create(HandLandmark.INDEX_FINGER_MCP, HandLandmark.INDEX_FINGER_PIP),
          Connection.create(HandLandmark.INDEX_FINGER_PIP, HandLandmark.INDEX_FINGER_DIP),
          Connection.create(HandLandmark.INDEX_FINGER_DIP, HandLandmark.INDEX_FINGER_TIP),
          Connection.create(HandLandmark.INDEX_FINGER_MCP, HandLandmark.MIDDLE_FINGER_MCP),
          Connection.create(HandLandmark.MIDDLE_FINGER_MCP, HandLandmark.MIDDLE_FINGER_PIP),
          Connection.create(HandLandmark.MIDDLE_FINGER_PIP, HandLandmark.MIDDLE_FINGER_DIP),
          Connection.create(HandLandmark.MIDDLE_FINGER_DIP, HandLandmark.MIDDLE_FINGER_TIP),
          Connection.create(HandLandmark.MIDDLE_FINGER_MCP, HandLandmark.RING_FINGER_MCP),
          Connection.create(HandLandmark.RING_FINGER_MCP, HandLandmark.RING_FINGER_PIP),
          Connection.create(HandLandmark.RING_FINGER_PIP, HandLandmark.RING_FINGER_DIP),
          Connection.create(HandLandmark.RING_FINGER_DIP, HandLandmark.RING_FINGER_TIP),
          Connection.create(HandLandmark.RING_FINGER_MCP, HandLandmark.PINKY_MCP),
          Connection.create(HandLandmark.WRIST, HandLandmark.PINKY_MCP),
          Connection.create(HandLandmark.PINKY_MCP, HandLandmark.PINKY_PIP),
          Connection.create(HandLandmark.PINKY_PIP, HandLandmark.PINKY_DIP),
          Connection.create(HandLandmark.PINKY_DIP, HandLandmark.PINKY_TIP));

  private static final String MODEL_COMPLEXITY = "model_complexity";
  private static final String NUM_HANDS = "num_hands";
  private static final String USE_PREV_LANDMARKS = "use_prev_landmarks";
  private static final String GPU_GRAPH_NAME = "hand_landmark_tracking_gpu_image.binarypb";
  private static final String CPU_GRAPH_NAME = "hand_landmark_tracking_cpu_image.binarypb";
  private static final String IMAGE_INPUT_STREAM = "image";
  private static final ImmutableList<String> OUTPUT_STREAMS =
      ImmutableList.of(
          "multi_hand_landmarks",
          "multi_hand_world_landmarks",
          "multi_handedness",
          "throttled_image");
  private static final int LANDMARKS_INDEX = 0;
  private static final int WORLD_LANDMARKS_INDEX = 1;
  private static final int HANDEDNESS_INDEX = 2;
  private static final int INPUT_IMAGE_INDEX = 3;
  private final OutputHandler<HandsResult> outputHandler;

  /**
   * Initializes MediaPipe Hands solution.
   *
   * @param context an Android {@link Context}.
   * @param options the configuration options defined in {@link HandsOptions}.
   */
  public Hands(Context context, HandsOptions options) {
    outputHandler = new OutputHandler<>();
    outputHandler.setOutputConverter(
        packets -> {
          HandsResult.Builder handsResultBuilder = HandsResult.builder();
          try {
            handsResultBuilder.setMultiHandLandmarks(
                getProtoVector(packets.get(LANDMARKS_INDEX), NormalizedLandmarkList.parser()));
          } catch (MediaPipeException e) {
            reportError("Error occurs while getting MediaPipe hand landmarks.", e);
          }
          try {
            handsResultBuilder.setMultiHandWorldLandmarks(
                getProtoVector(packets.get(WORLD_LANDMARKS_INDEX), LandmarkList.parser()));
          } catch (MediaPipeException e) {
            reportError("Error occurs while getting MediaPipe hand world landmarks.", e);
          }
          try {
            List<Classification> handednessList = new ArrayList<>();
            for (ClassificationList protolist :
                getProtoVector(packets.get(HANDEDNESS_INDEX), ClassificationList.parser())) {
              handednessList.add(protolist.getClassification(0));
            }
            handsResultBuilder.setMultiHandedness(handednessList);
          } catch (MediaPipeException e) {
            reportError("Error occurs while getting MediaPipe handedness data.", e);
          }
          return handsResultBuilder
              .setImagePacket(packets.get(INPUT_IMAGE_INDEX))
              .setTimestamp(
                  staticImageMode ? Long.MIN_VALUE : packets.get(INPUT_IMAGE_INDEX).getTimestamp())
              .build();
        });

    SolutionInfo solutionInfo =
        SolutionInfo.builder()
            .setBinaryGraphPath(options.runOnGpu() ? GPU_GRAPH_NAME : CPU_GRAPH_NAME)
            .setImageInputStreamName(IMAGE_INPUT_STREAM)
            .setOutputStreamNames(OUTPUT_STREAMS)
            .setStaticImageMode(options.staticImageMode())
            .build();

    initialize(context, solutionInfo, outputHandler);
    Map<String, Packet> inputSidePackets = new HashMap<>();
    inputSidePackets.put(NUM_HANDS, packetCreator.createInt32(options.maxNumHands()));
    inputSidePackets.put(MODEL_COMPLEXITY, packetCreator.createInt32(options.modelComplexity()));
    inputSidePackets.put(USE_PREV_LANDMARKS, packetCreator.createBool(!options.staticImageMode()));
    start(inputSidePackets);
  }

  /**
   * Sets a callback to be invoked when a {@link HandsResult} becomes available.
   *
   * @param listener the {@link ResultListener} callback.
   */
  public void setResultListener(ResultListener<HandsResult> listener) {
    this.outputHandler.setResultListener(listener);
  }

  /**
   * Sets a callback to be invoked when the Hands solution throws errors.
   *
   * @param listener the {@link ErrorListener} callback.
   */
  public void setErrorListener(@Nullable ErrorListener listener) {
    this.outputHandler.setErrorListener(listener);
    this.errorListener = listener;
  }
}
